libfwupdplugin/fu-firmware-attributes.c

```diff
@@ -0,0 +1,156 @@
+/*
+ * Copyright (C) 2022 Mario Limonciello <mario.limonciello@amd.com>
+ *
+ * SPDX-License-Identifier: LGPL-2.1+
+ */
+
+#define G_LOG_DOMAIN "FuFirmwareAttributes"
+
+#include "config.h"
+
+#include "fu-firmware-attributes.h"
+#include "fu-path.h"
+#include "fu-string.h"
+
+static gchar *
+fu_firmware_attributes_build(const gchar *name, GError **error)
+{
+	g_autofree gchar *sysfsfwdir = fu_path_from_kind(FU_PATH_KIND_SYSFSDIR_FW_ATTRIB);
+	g_autofree gchar *tmp = NULL;
+	const gchar *current_value;
+	const gchar *driver;
+	GDir *dir;
+
+	g_return_val_if_fail(name != NULL, NULL);
+
+	dir = g_dir_open(sysfsfwdir, 0, error);
+	if (dir == NULL)
+		return NULL;
+
+	driver = g_dir_read_name(dir);
+	if (driver == NULL) {
+		g_set_error(error,
+			    G_IO_ERROR,
+			    G_IO_ERROR_NOT_FOUND,
+			    "failed to find driver in %s",
+			    sysfsfwdir);
+		return NULL;
+	}
+
+	if (g_dir_read_name(dir)) {
+		g_set_error(error,
+			    G_IO_ERROR,
+			    G_IO_ERROR_NOT_FOUND,
+			    "Unable to support multiple drivers found in %s",
+			    sysfsfwdir);
+		return NULL;
+	}
+
+	current_value = g_strcmp0(name, "pending_reboot") != 0 ? "current_value" : NULL;
+	tmp = g_build_filename(sysfsfwdir, driver, "attributes", name, current_value, NULL);
+	if (!g_file_test(tmp, G_FILE_TEST_EXISTS)) {
+		g_set_error(error,
+			    G_IO_ERROR,
+			    G_IO_ERROR_NOT_FOUND,
+			    "failed to find attribute %s",
+			    tmp);
+		return NULL;
+	}
+
+	return g_steal_pointer(&tmp);
+}
+
+/**
+ * fu_firmware_attribute_has_attr:
+ * @name: A firmware attribute name found in /sys/class/firmware-attributes/$DRIVER/attributes
+ *
+ * Determines if the system offers a given firmware attribute name
+ *
+ * Since: 1.8.4
+ **/
+gboolean
+fu_firmware_attribute_has_attr(const gchar *name)
+{
+	g_autofree gchar *tmp = fu_firmware_attributes_build(name, NULL);
+	return (tmp != NULL);
+}
+
+/**
+ * fu_firmware_attributes_get_value:
+ * @name: A firmware attribute name found in /sys/class/firmware-attributes/$DRIVER/attributes
+ * @error: (nullable): optional return location for an error
+ *
+ * Gets the string stored in an attribute
+ *
+ * Since: 1.8.4
+ **/
+gchar *
+fu_firmware_attributes_get_value(const gchar *name, GError **error)
+{
+	g_autofree gchar *path = fu_firmware_attributes_build(name, error);
+	g_autofree gchar *buf = NULL;
+
+	if (path == NULL)
+		return NULL;
+
+	if (!g_file_get_contents(path, &buf, NULL, error)) {
+		g_prefix_error(error, "failed to read attribute from %s: ", path);
+		return NULL;
+	}
+
+	return g_steal_pointer(&buf);
```

Comment:
is this always going to be UTF-8 text? I think renaming this to fu_firmware_attributes_get_string() and using fu_strsafe() with the length is a good idea, and also providing a fu_firmware_attributes_get_bytes() that returns a binary blob.

Reply:
I found that it needs to be stripped first or fu_strsafe causes problems.  It also doesn't work for reading a single integer.

Reply:
Okay gotcha. I still think we need to check that the GBytes has a trailing NUL, no? i.e. a g_strndup() somewhere.